kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.snapshot;
+
+import static io.delta.kernel.internal.DeltaErrors.wrapEngineExceptionThrowsIO;
+import static io.delta.kernel.internal.checkpoints.Checkpointer.getLatestCompleteCheckpointFromList;
+import static io.delta.kernel.internal.lang.ListUtils.getFirst;
+import static io.delta.kernel.internal.lang.ListUtils.getLast;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.stream.Collectors.toList;
+
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.checkpoints.CheckpointInstance;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.util.Clock;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MetadataCleanup {
+  private static final Logger logger = LoggerFactory.getLogger(MetadataCleanup.class);
+
+  public MetadataCleanup() {}
+
+  /**
+   * Delete the Delta log files (delta and checkpoint files) that are expired according the table
+   * metadata retention settings. While deleting the log files, it makes sure the time travel
+   * continues to work for all unexpired table versions.
+   *
+   * <p>Here is algorithm:
+   *
+   * <ul>
+   *   <li>Initial the potential delete file list: `potentialFilesToDelete` as empty list
+   *   <li>Initialize the last seen checkpoint file list: `lastSeenCheckpointFiles`. There could be
+   *       one or more checkpoint files for a given version.
+   *   <li>List the delta log files starting with prefix "00000000000000000000." (%020d). For each
+   *       file:
+   *       <ul>
+   *         <li>Step 1: Check if the `lastSeenCheckpointFiles` contains a complete checkpoint, then
+   *             <ul>
+   *               <li>Step 1.1: delete all files in `potentialFilesToDelete`. Now we know there is
+   *                   a checkpoint that contains the compacted Delta log up to the checkpoint
+   *                   version and all commit/checkpoint files before this checkpoint version are
+   *                   not needed.
+   *               <li>Step 1.2: add `lastCheckpointFiles` to `potentialFileStoDelete` list. This
+   *                   checkpoint is potential candidate to delete later if we find another
+   *                   checkpoint
+   *             </ul>
+   *         <li>Step 2: If the timestamp is earlier than the retention period, stop
```

Comment:
let's not say "earlier", that's confusing, let's say "newer".

Reply:
"earlier" is better word here as we are talking about time.

Reply:
changed it to falls within the retention period.